ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -121,4 +107,97 @@ public void start() throws ReplicationException {
       logger.error("IO error while starting replication", e);
     }
   }
+
+  /**
+   * Add given replica into replication manager.
+   * @param replicaId the replica to add
+   * @return {@code true} if addition succeeded, {@code false} failed to add replica because it already exists.
+   */
+  public boolean addReplica(ReplicaId replicaId) {
+    if (partitionToPartitionInfo.containsKey(replicaId.getPartitionId())) {
+      logger.error("{} already exists in replication manager, rejecting adding replica request.",
+          replicaId.getPartitionId());
+      return false;
+    }
+    List<? extends ReplicaId> peerReplicas = replicaId.getPeerReplicaIds();
+    List<RemoteReplicaInfo> remoteReplicaInfos = new ArrayList<>();
+    if (peerReplicas != null) {
+      remoteReplicaInfos = createRemoteReplicaInfos(peerReplicas, replicaId);
+      updatePartitionInfoMaps(remoteReplicaInfos, replicaId);
+    }
+    logger.info("Assigning thread for {}", replicaId.getPartitionId());
+    addRemoteReplicaInfoToReplicaThread(remoteReplicaInfos, true);
+    // No need to update persistor to explicitly persist tokens for new replica because background persistor will
+    // periodically persistor all tokens including new added replica's
+    // TODO ensure there is a test that checks persistor has new replica info.
+    logger.info("{} is successfully added into replication manager", replicaId.getPartitionId());
+    return true;
+  }
+
+  /**
+   * Remove replica from replication manager
+   * @param replicaId the replica to remove
+   * @return {@code true} if replica is successfully removed. {@code false} otherwise
+   */
+  public boolean removeReplica(ReplicaId replicaId) {
+    if (!partitionToPartitionInfo.containsKey(replicaId.getPartitionId())) {
+      logger.error("{} doesn't exist in replication manager, skipping removing replica request.",
+          replicaId.getPartitionId());
+      return false;
```

Comment:
Can the add/remove replica code cause thread safety concerns ? Let's discuss! There are some subtle differences between ConcurrentMap and ConcurrentHashMap and we should try to explicitly use ConcurrentHashMap is we want certain guarantees.

Reply:
Good point. I decided to add `ReadWriteLock` in replication manager and disk manager.